File: solutions/php/bank-account/1/BankAccount.php
<?php
declare(strict_types=1);

class BankAccount
{
    private int $balance = 0;
    private bool $isOpen = false;

    public function open()
    {
        if ($this->isOpen) {
            throw new InvalidArgumentException('account already open');
        }
        $this->isOpen = true;
        $this->balance = 0;
    }

    public function close()
    {
        if (!$this->isOpen) {
            throw new InvalidArgumentException('account not open');
        }
        $this->isOpen = false;
    }

    public function balance(): int
    {
        if (!$this->isOpen) {
            throw new InvalidArgumentException('account not open');
        }
        return $this->balance;
    }

    public function deposit(int $amt)
    {
        if (!$this->isOpen) {
            throw new InvalidArgumentException('account not open');
        }
        if ($amt < 0) {
            throw new InvalidArgumentException('amount must be greater than 0');
        }
        $this->balance += $amt;
    }

    public function withdraw(int $amt)
    {
        if (!$this->isOpen) {
            throw new InvalidArgumentException('account not open');
        }
        if ($amt < 0) {
            throw new InvalidArgumentException('amount must be greater than 0');
        }
        if ($amt > $this->balance) {
            throw new InvalidArgumentException(
                'amount must be less than balance',
            );
        }
        $this->balance -= $amt;
    }
}

File: solutions/lua/bank-account/2/bank-account.lua
local BankAccount = {}

function BankAccount:new()
    return setmetatable({ _balance = 0, _isOpen = false }, { __index = self })
end

function BankAccount:open()
    assert(not self._isOpen)
    self._balance = 0
    self._isOpen = true
end

function BankAccount:close()
    assert(self._isOpen)
    self._isOpen = false
end

function BankAccount:deposit(amount)
    assert(self._isOpen)
    assert(amount >= 0)
    self._balance = self._balance + amount
end

function BankAccount:withdraw(amount)
    assert(self._isOpen)
    assert(amount >= 0)
    assert(amount <= self._balance)
    self._balance = self._balance - amount
end

function BankAccount:balance()
    assert(self._isOpen)
    return self._balance
end


return BankAccount
